demo/llama/train.py

```diff
@@ -0,0 +1,389 @@
+
+        input_ids = train_data[:, 0:model.config.block_size].contiguous()
+        targets = train_data[:, 1:model.config.block_size + 1].contiguous()
+
+        is_accumulating = (iter_num + 1) % grad_accum_steps != 0
+
+        with fabric.no_backward_sync(model, enabled=is_accumulating):
+            logits = model(input_ids)
+            loss = torch.nn.functional.cross_entropy(logits.view(
+                -1, logits.size(-1)),
+                                                     targets.view(-1),
+                                                     ignore_index=-1)
+            fabric.backward(loss / grad_accum_steps)
+
+        t1 = time.time()
+
+        if not is_accumulating:
+            fabric.clip_gradients(model, optimizer, max_norm=grad_clip)
+
+            optimizer.step()
+            optimizer.zero_grad()
+            step_count += 1
+
+            t1 = time.time()
+
+            if val_dataloader is not None and step_count % eval_interval == 0:
+                val_loss = validate(fabric, model, val_dataloader)
+                fabric.print(f"step {iter_num}: val loss {val_loss:.4f}")
+                fabric.barrier()
+                fabric.log_dict({
+                    "iter": iter_num,
+                    "val_loss": val_loss,
+                    "step": step_count,
+                    "lr": lr
+                })
+
+            if step_count % save_interval == 0:
+                cur_ckpt_save_dir = os.path.join(checkpoint_save_dir,
+                                                 f"iter-{iter_num:06d}")
+                fabric.print(f"Saving checkpoint to {cur_ckpt_save_dir}")
+                state = {
+                    "model": model,
+                    "optimizer": optimizer,
+                    "iteration": iter_num
+                }
+                save_start = time.time()
+                fabric.save(cur_ckpt_save_dir, state)
+                save_end = time.time()
+
+                if fabric.global_rank == 0:
+                    fabric.print(
+                        f"Checkpoint save with dataflux took {save_end - save_start} seconds."
+                    )
+
+        dt = t1 - t0
+
+        tokens += micro_batch_size * model.config.block_size
+        step_time += t1 - prev_t1
+        prev_t1 = t1
+
+        if iter_num % log_interval == 0:
+            tokens_sec_str = f"{tokens / step_time:.0f}" if not is_accumulating else "-"
+
+            fabric.log_dict({
+                "iter": iter_num,
+                "train_loss": loss,
+                "step": step_count,
+                "lr": lr
+            })
+            fabric.print(
+                f"iter {iter_num}: loss {loss.item():.4f}, time: {dt*1000:.2f}ms, speed: {tokens_sec_str} toks/s/device"
+            )
+
+        if not is_accumulating:
+            tokens = 0
+            step_time = 0.0
+
+        if iter_num > max_iters:
+            break
+
+
+@torch.no_grad()
+def validate(fabric: L.Fabric, model: torch.nn.Module,
```

Comment:
Same comment here and for several of the functions below, does this really need to be reimplemented?